bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -25,219 +24,396 @@ import jsky.app.ot.gemini.tpe.TpePWFSFeature
-    def +(key: SPNodeKey): BagsState = copy(keys = keys + key)
-    def -(key: SPNodeKey): BagsState = copy(keys = keys - key)
-    def +(pid: SPProgramID): BagsState = copy(programs = programs + pid)
-    def -(pid: SPProgramID): BagsState = copy(programs = programs - pid)
-  }
-  @volatile private var state: BagsState = BagsState(Set.empty, Set.empty)
+  /** Called to notify that an AGS lookup failed. */
+  def fail(why: String, delayMs: Long): StateTransition =
+    ErrorTransition
+}
 
-  // Setup an LRU cache with max entries BagsManager.CacheSize
-  private val hashes = new util.LinkedHashMap[SPNodeKey, Int](math.ceil((BagsManager.CacheSize + 1) * 1.334).toInt, 0.75f, true) {
-    override def removeEldestEntry(e: java.util.Map.Entry[SPNodeKey, Int]): Boolean =
-      size > BagsManager.CacheSize
+
+object BagsState {
+  type AgsHashVal      = Long
+  type StateTransition = (BagsState, IO[Unit]) // Next BagsState, side-effects
+  val ErrorTransition  = (ErrorState, ioUnit)  // Default transition for unexpected events
+
+  /** Error.  This state represents a logic error.  There are no transitions out
+    * of the error state.
+    */
+  case object ErrorState extends BagsState
+
+  /** Idle.  When idle, there are no pending or running AGS look-ups.  This is
+    * where we hang out waiting for something to be edited in order to kick off
+    * another AGS search.  We hang on to the last known AGS hash value so that
+    * we can discard irrelevant edits if possible.
+    */
+  case class IdleState(obs: ISPObservation, hash: Option[AgsHashVal]) extends BagsState {
+    override val edit: StateTransition =
+      (PendingState(obs, hash), BagsManager.wakeUpAction(obs, 0))
   }
 
-  /**
-    * Atomically add a program to our watch list and attach listeners.
-    * Enqueue all observations for consideration for a BAGS lookup.
+  /** Pending.  When pending, we assume we've been edited and that a new AGS
+    * search may be required.  This is a transient state since we're expecting
+    * the timer to go off, wake us up and then we'll decide whether to do an
+    * AGS lookup, update the AGS hash code, etc.
     */
-  def watch(prog: ISPProgram): Unit = {
-    synchronized {
-      state += prog.getProgramID
-      prog.addStructureChangeListener(StructurePropertyChangeListener)
-      prog.addCompositeChangeListener(CompositePropertyChangeListener)
+  case class PendingState(obs: ISPObservation, hash: Option[AgsHashVal]) extends BagsState {
+    override val edit: StateTransition =
+      (this, ioUnit)
+
+    override def wakeUp: StateTransition = {
+      def notObserved: Boolean =
+        ObservationStatus.computeFor(obs) != ObservationStatus.OBSERVED
+
+      // Get the ObsContext and AgsStrategy, if possible.  If not possible,
+      // we won't be able to do AGS for this observation.
+      val tup = for {
+        c <- ObsContext.create(obs).asScalaOpt
+        s <- agsStrategy(obs, c)
+      } yield (c, s)
```

Comment:
Do we even need `s` here? It looks like it is never used from what I can tell.
I think we could get away with replacing `tup` with an `Option[ObsContext]` and do so only `if agsStrategy(obs,c).isDefined`, or just change the `agsStrategy` line to `_ <- agsStrategy(obs, c)` and then `yield c`.